elastic/kibana
# [Security Solution] Change query builder so N exception items don't nest N levels deep

## Summary

Currently, the usage of buildEsQuery with a single large KQL string causes a deeply nested Elasticsearch query to be generated. The builder generates an AST from a boolean expression of the form `A OR B OR C OR D ...`, and rather than detecting that A, B, C, and D can be put in an array in a single `should` block, it creates a set of nested `should` blocks as deep as the number of expressions being OR'd together. In our case, the number of expressions being OR'd is the number of exception items. This eventually causes Elasticsearch to fail on the search when the nesting becomes too deep.

This PR changes the logic so that every exception item being OR'd together will be in the same array without unnecessary nesting.  A further improvement would be using similar logic on the entries within an exception, which can suffer from the same problem with the AST, however the number of entries per exception has a much lower practical limit.

Additional refactoring for next PR:
- [ ] Remove unnecessary `language` option from `buildExceptionListQueries`
- [x] Extract building of `exceptionFilter` to separate function
- [ ] Handle potential error case if maxClauseCount is below 1024 for some reason
- [ ] Potentially increase the exception list size limit from 5000 https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts#L186

Manual test process:
- Add ~4000 endpoint exception list items by modifying [this script](https://github.com/elastic/kibana/blob/master/x-pack/plugins/lists/server/scripts/post_endpoint_list_item.sh)
- Create a detection alert
- Add an exception that allowlists the detection alert
- Observe that no more detection alerts are created that match the new exception

### For maintainers
